File: logging middleware/NotificationPrioritySystem.py
import heapq
from datetime import datetime

class Notification:
    WEIGHTS = {
        "placement": 3,
        "result": 2,
        "event": 1
    }

    def __init__(self, notification_id, category, timestamp, is_read=False):
        self.notification_id = notification_id
        self.category = category.lower()
        self.timestamp = timestamp
        self.is_read = is_read

    def priority_score(self):
        weight = self.WEIGHTS.get(self.category, 0)
        return weight * 10**12 + self.timestamp


class NotificationSystem:
    def __init__(self):
        self.heap = []

    def add_notification(self, notification):
        if not notification.is_read:
            score = notification.priority_score()
            heapq.heappush(self.heap, (-score, notification))

    def get_top_notifications(self, n=10):
        top = heapq.nsmallest(n, self.heap)
        return [item[1] for item in top]


if __name__ == "__main__":
    system = NotificationSystem()

    now = int(datetime.now().timestamp())

    system.add_notification(Notification("N1", "placement", now - 10))
    system.add_notification(Notification("N2", "result", now - 20))
    system.add_notification(Notification("N3", "event", now - 30))
    system.add_notification(Notification("N4", "placement", now - 40))
    system.add_notification(Notification("N5", "result", now - 50))
    system.add_notification(Notification("N6", "placement", now - 60))
    system.add_notification(Notification("N7", "event", now - 70))
    system.add_notification(Notification("N8", "placement", now - 5))
    system.add_notification(Notification("N9", "result", now - 15))
    system.add_notification(Notification("N10", "event", now - 25))
    system.add_notification(Notification("N11", "placement", now))

    top_notifications = system.get_top_notifications(10)

    print("Top 10 Unread Notifications:")
    for n in top_notifications:
        print(
            f"ID={n.notification_id}, "
            f"Category={n.category}, "
            f"Timestamp={n.timestamp}"
        )

File: logging middleware/NotificationPrioritySystem.java
import java.util.*;

class Notification {
    String id;
    String type; // Placement, Result, Event
    long timestamp;
    boolean isRead;

    public Notification(String id, String type, long timestamp, boolean isRead) {
        this.id = id;
        this.type = type;
        this.timestamp = timestamp;
        this.isRead = isRead;
    }

    public int getWeight() {
        switch (type.toLowerCase()) {
            case "placement":
                return 3;
            case "result":
                return 2;
            case "event":
                return 1;
            default:
                return 0;
        }
    }

    public double getPriorityScore() {
        return getWeight() * 1_000_000_000.0 + timestamp;
    }

    @Override
    public String toString() {
        return "Notification{" +
                "id='" + id + '\'' +
                ", type='" + type + '\'' +
                ", timestamp=" + timestamp +
                '}';
    }
}

public class NotificationPrioritySystem {

    private final PriorityQueue<Notification> heap;

    public NotificationPrioritySystem() {
        heap = new PriorityQueue<>(
                (a, b) -> Double.compare(
                        b.getPriorityScore(),
                        a.getPriorityScore()
                )
        );
    }

    public void addNotification(Notification notification) {
        if (!notification.isRead) {
            heap.offer(notification);
        }
    }

    public List<Notification> getTopNotifications(int n) {
        List<Notification> result = new ArrayList<>();
        List<Notification> temp = new ArrayList<>();

        int count = 0;

        while (!heap.isEmpty() && count < n) {
            Notification current = heap.poll();
            result.add(current);
            temp.add(current);
            count++;
        }

        heap.addAll(temp);

        return result;
    }

    public static void main(String[] args) {

        NotificationPrioritySystem system = new NotificationPrioritySystem();

        long now = System.currentTimeMillis();

        system.addNotification(new Notification("N1", "Placement", now - 1000, false));
        system.addNotification(new Notification("N2", "Result", now - 2000, false));
        system.addNotification(new Notification("N3", "Event", now - 3000, false));
        system.addNotification(new Notification("N4", "Placement", now - 4000, false));
        system.addNotification(new Notification("N5", "Result", now - 5000, false));
        system.addNotification(new Notification("N6", "Placement", now - 6000, false));
        system.addNotification(new Notification("N7", "Event", now - 7000, false));
        system.addNotification(new Notification("N8", "Result", now - 8000, false));
        system.addNotification(new Notification("N9", "Placement", now - 9000, false));
        system.addNotification(new Notification("N10", "Event", now - 10000, false));
        system.addNotification(new Notification("N11", "Placement", now, false));

        List<Notification> top10 = system.getTopNotifications(10);

        System.out.println("Top 10 Notifications:");
        for (Notification n : top10) {
            System.out.println(n);
        }
    }
}